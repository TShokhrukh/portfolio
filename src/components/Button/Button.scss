
.Button {
  cursor: pointer;
  border: 1px solid $purple;
  padding: $spacer * 1.25 $spacer * 2;
  border-radius: $border-radius;
  background: none;
  color: $purple;
  display: flex;
  align-items: center;
}

.Button_disabled {
  cursor: default;
}

.Button_rounded {
  border-radius: 50%;
}
