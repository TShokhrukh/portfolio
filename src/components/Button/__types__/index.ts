import type { ElementType, ComponentType, ReactNode, SyntheticEvent } from 'react'
import type { IClassNameProps } from '@bem-react/core'

export interface IIconProps {
  className: string
}

export interface IButtonProps extends IClassNameProps {
  component?: ElementType,
  startIcon?: ComponentType<IIconProps>,
  endIcon?: ComponentType<IIconProps>,
  rounded?: boolean,
  disabled?: boolean,
  type?: 'submit'|'button'|'reset',
  id?: string,
  children?: ReactNode,
  onClick?: (event: SyntheticEvent<HTMLButtonElement>) => void
}

export interface IButtonVariantLinkProps {
  variant?: 'link',
  type?: undefined,
  href?: string
}

export interface IButtonIconProps {
  icon: ComponentType<IIconProps>,
  direction: 'start'|'end'
}
