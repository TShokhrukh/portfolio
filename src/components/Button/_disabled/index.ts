export * from './Button_disabled'
