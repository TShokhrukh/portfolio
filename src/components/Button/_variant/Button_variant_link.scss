
.Button_variant_link {

}
