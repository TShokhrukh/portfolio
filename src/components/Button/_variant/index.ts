export * from './Button_variant_link'
