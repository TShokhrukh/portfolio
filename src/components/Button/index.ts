export * from './build'
export * from './__types__'
export * from './_disabled'
export * from './_rounded'
export * from './_variant'
export { Button as ButtonBase } from './Button'
