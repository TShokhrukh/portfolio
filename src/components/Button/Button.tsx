import React, { FC } from 'react'

import { ButtonIcon } from './Icon/Button-Icon'
import { cnButton } from './Button.const'
import type { IButtonProps } from './__types__'

import './Button.scss'

export const Button: FC<IButtonProps> = ({
  component: Component = 'button',
  type = 'button',
  children,
  className,
  disabled,
  rounded,
  startIcon,
  endIcon,
  ...props
}) => (
  <Component {...props}
    className={cnButton({ disabled, rounded }, [className])}
    type={type}
    disabled={disabled}>
    {startIcon && <ButtonIcon icon={startIcon} direction='start' />}
    {children}
    {endIcon && <ButtonIcon icon={endIcon} direction='end' />}
  </Component>
)
