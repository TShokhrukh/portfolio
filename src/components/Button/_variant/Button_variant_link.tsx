import React from 'react'
import { withBemMod } from '@bem-react/core'
import { cnButton } from '../Button.const'
import type { IButtonProps, IButtonVariantLinkProps } from '../__types__'
import './Button_variant_link.scss'

export const withButtonVariantLink = withBemMod<IButtonVariantLinkProps, IButtonProps>(
  cnButton(),
  { variant: 'link' },
  (Button) => function ButtonTypeLink (props) {
    return <Button {...props} component='a' />
  }
)
