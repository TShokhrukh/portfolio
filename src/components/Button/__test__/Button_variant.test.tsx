/* global test, expect, describe */
import React from 'react'
import { compose } from '@bem-react/core'
import { render, screen } from '@testing-library/react'
import { Button as ButtonBase } from '../Button'
import { withButtonVariantLink } from '../_variant/Button_variant_link'

const Button = compose(withButtonVariantLink)(ButtonBase)

describe('Button_variant', () => {
  test('component variant link', () => {
    render(<Button data-testid='button' variant='link' href='/home'>children</Button>)

    expect(screen.getByTestId('button')).toMatchSnapshot()
  })
})
