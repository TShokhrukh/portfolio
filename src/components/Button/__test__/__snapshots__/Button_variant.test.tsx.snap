// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button_variant component variant link 1`] = `
<a
  class="Button Button_variant_link"
  data-testid="button"
  href="/home"
  type="button"
  variant="link"
>
  children
</a>
`;
